qemu/tests/smt_test.py

```diff
@@ -0,0 +1,119 @@
+from virttest import utils_misc
+from virttest import utils_test
+
+from virttest.utils_test import BackgroundTest
+from provider.cpu_utils import check_cpu_flags
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    smt test:
+    1) Check if host has topoext flag, if not, cancel test
+    2) Boot guest and check cpu count and threads number in guest
+    3) Run stress inside guest, and check cpu usage in guest(only for linux)
+
+    :params test: QEMU test object.
+    :params params: Dictionary with the test parameters.
+    :params env: Dictionary with test environment.
+    """
+    def get_guest_threads():
+        """
+        Get guest threads number
+        """
+        if os_type == "linux":
+            cmd = params["get_threads_cmd"]
+            output = session.cmd_output_safe(cmd)
+            threads = int(re.findall(r":\s*(\d+)", output)[0])
+        else:
+            cmd = params["get_cores_cmd"]
+            output = session.cmd_output_safe(cmd)
+            cores = int(re.findall(r"=(\d+)", output)[0])
+            cmd = params["get_sockets_cmd"]
+            output = session.cmd_output_safe(cmd)
+            sockets = len(re.findall(r"SocketDesignation=", output))
+            threads = int(vm.cpuinfo.smp/sockets/cores)
+        return threads
+
+    def heavyload_install(install_path):
+        """
+        Install heavyload in windows guest
+        """
+        test_installed_cmd = 'dir "%s" | findstr /I heavyload' % install_path
+        if session.cmd_status(test_installed_cmd) != 0:
+            logging.warning("Could not find installed heavyload in guest, will"
+                            " install it via winutils.iso ")
+            winutil_drive = utils_misc.get_winutils_vol(session)
+            if not winutil_drive:
+                test.cancel("WIN_UTILS CDROM not found.")
+            install_cmd = params["install_cmd"] % winutil_drive
+            session.cmd(install_cmd)
+
+    def run_stress():
+        """
+        Run stress inside guest, return guest cpu usage
+        """
+        error_context.context("Run stress in guest and get cpu usage",
+                              logging.info)
+        if os_type == "linux":
+            stress_args = params["stress_args"]
+            stress_test = utils_test.VMStress(vm, "stress",
+                                              params, stress_args=stress_args)
+            try:
+                stress_test.load_stress_tool()
+                time.sleep(stress_duration / 2)
+                output = session.cmd_output_safe(params["get_cpu_usage_cmd"])
+                utils_misc.wait_for(lambda: (stress_test.app_running is False), 30)
+                stress_test.unload_stress()
+                cpu_usage = re.findall(r":\s*(\d+.?\d+)\s*us", output)
+                cpu_usage = [float(x) for x in cpu_usage]
+                logging.info("Guest cpu usage is %s", cpu_usage)
+                unloaded_cpu = [x for x in cpu_usage if x < 20]
```

Comment:
Would you help illustrate why `20` here?

Reply:
Normally, if a cpu is not loaded, the usage should be 0.  In case it's loaded by other tasks other than stress, a bigger number should be used. As I observed, when stress runs, the usage for each cpu varies from 40 ~ 100. To give a tolerate range, I set 20 as the threshold. Does it make sense to you?

Reply:
@huangyum I guess it depends on the stress argument you set and the cpu configuration of the guest, both of them are set in this case, so it should be stable and no risk for missing bug. So yes it's okay for me. Thanks.